pkg/cmd/pulumi/plugin_install_test.go

```diff
@@ -0,0 +1,44 @@
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/env"
+	"github.com/stretchr/testify/assert"
+)
+
+// Test for https://github.com/pulumi/pulumi/issues/11703, check we give an error when trying to install a
+// bundled plugin
+func TestBundledError(t *testing.T) {
+	t.Parallel()
+
+	cmd := &pluginInstallCmd{
+		e: env.NewEnv(
+			env.MapStore{"PULUMI_DEV": "false"},
+		),
+	}
+
+	err := cmd.Run(context.Background(), []string{"language", "nodejs"})
+	assert.EqualError(t, err,
+		"the nodejs language plugin is bundled with Pulumi, "+
+			"and cannot be directly installed with this command. "+
+			"If you need to reinstall this plugin, reinstall Pulumi via your package manager or install script.")
+}
+
+// Test for https://github.com/pulumi/pulumi/issues/11703, check we still try to install bundled plugins if
+// PULUMI_DEV is set.
+func TestBundledDev(t *testing.T) {
+	t.Parallel()
+
+	cmd := &pluginInstallCmd{
+		e: env.NewEnv(
+			env.MapStore{"PULUMI_DEV": "true"},
+		),
+	}
+
+	// TODO: This test actually hits GitHub to fail, it would be better if plugin installation could be mocked.
```

Comment:
cc @abhinav your generally good at thinking up clean designs for this sort of thing, any recommendation here.
I've been using functions for this in some places (e.g. in pkg/codegen/convert/plugin_mapper.go), maybe not overkill to do that here?

Reply:
I think a function reference to workspace.DownloadToFile or similar is fine here.
Longer term, it makes sense for these top-level functions with so much complex behavior to be turned into objects with clearer contracts that we inject in here, e.g. a `*PluginDownloader` struct and an interface for that in the command implementation. Then we can mock it out and return an error on a download attempt, or start up a test HTTP server and make the downloader implementation hit that.